File: 0042-trapping-rain-water/0042-trapping-rain-water.c


int trap(int* height, int heightSize)
{
    int heightL[heightSize];
    int heightR[heightSize];
    for(int i = 0; i < heightSize; i++)
    {
        heightL[i] = 0;
        heightR[i] = 0;
    }

    for(int i = 1; i < heightSize; i++)
    {
        heightL[i] = max(heightL[i - 1], height[i - 1]);
        heightR[heightSize - i - 1] = max(heightR[heightSize - i], height[heightSize - i]);
    }
    int x = 0;
    for(int i = 0; i < heightSize; i++)
    {
        x += max(0, min(heightL[i], heightR[i]) - height[i]);
    }
    return x;
};

int max(int a, int b)
{
    return (a > b) ? a : b;
};

int min(int a, int b)
{
    return (a < b) ? a : b;
};

File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height):
        heightR = [0 for _ in range(len(height))]
        heightL = [0 for _ in range(len(height))]
        for i in range(1, len(height)):
            heightL[i] = max(heightL[i - 1], height[i - 1])
            heightR[~i] = max(heightR[-i], height[-i])
        x = 0
        for i in range(len(height)):
            x += max(0, min(heightL[i], heightR[i]) - height[i])
        return x